FIAT/discontinuous_lagrange.py

```diff
@@ -5,7 +5,110 @@
 from FIAT import finite_element, polynomial_set, dual_set, functional, P0
+from FIAT.finite_element import make_barycentric_lattice_coordinates, make_index_permutations
+
+
+def make_entity_permutations(dim, npoints):
+    if npoints <= 0:
+        return {o: [] for o in range(np.math.factorial(dim + 1))}
+    # DG nodes are numbered, in order of significance,
+    # - by g0: entity dim (vertices first, then edges, then ...)
+    # - by g1: entity ids (DoFs on entities of smaller ids first)
+    # - lexicographically as in CG
+    #
+    # Example:
+    # dim = 2, degree = 3 (npoints = degree + 1)
+    #
+    #     facet ids    CG node numbers  DG node numbers
+    #    +
+    #    | \           3                1
+    #    |   \  0      2 6              8 3
+    #  2 |     \       1 5 8            7 9 4
+    #    |       \     0 4 7 9          0 5 6 2
+    #    +--------+
+    #        1
+    a = make_barycentric_lattice_coordinates(dim, npoints - 1)
+    # a = [[0, 0, 3],
+    #      [0, 1, 2],    (0,3,0)
+    #      [0, 2, 1],
+    #      [0, 3, 0],    (0,2,1)(1,2,0)
+    #      [1, 0, 2],
+    #      [1, 1, 1],    (0,1,2)(1,1,1)(2,1,0)
+    #      [1, 2, 0],
+    #      [2, 0, 1],    (0,0,3)(1,0,2)(2,0,1)(3,0,0)
+    #      [2, 1, 0],
+    #      [3, 0, 0]]    Lattice points that a represents
+    # a.shape[0] = number of DoFs
+    # a.shape[1] = dim + 1
+    # sum of each row = degree (bary centric lattice coordinates)
+    index_perms = make_index_permutations(dim + 1)
+    # index_perms = [[0, 1, 2],
+    #                [0, 2, 1],
+    #                [1, 0, 2],
+    #                [1, 2, 0],
+    #                [2, 0, 1],
+    #                [2, 1, 0]]
+    g0 = dim - (a == 0).astype(int).sum(axis=1)
+    # g0 = [ 0, 1, 1, 0, 1, 2, 1, 1, 1, 0]  # 0 for vertices
+    #                                       # 1 for edges
+    #                                       # 2 for cell
+    g1 = np.zeros_like(g0)
+    for d in range(dim + 1):
+        on_facet_d = (a[:, d] == 0).astype(int)
+        g1 += d * on_facet_d
+    # The above logic is consistent with the FIAT entity numbering
+    # convention ("entities that are not incident to vertex 0 are
+    # numbered first, then ..."), but vertices are not numbered using
+    # the same logic in FIAT, so we need to fix:
+    g1[g0 == 0] = -g1[g0 == 0]
+    # g1 = [-3, 2, 2,-2, 1, 0, 0, 1, 0,-1]
+    # For each entity dimension, DoFs that have smaller numbers in g1
+    # will be assigned smaller DG node numbers.
+    # Order first by g0, then by g1, and finally by a (lexicographically as in CG)
+    g0 = g0.reshape((a.shape[0], 1))
+    g1 = g1.reshape((a.shape[0], 1))
+    dg_cg_map = np.lexsort(np.transpose(np.concatenate((a, g1, g0), axis=1)))
+    # dg_cg_map = [ 0, 3, 9, 6, 8, 4, 7, 1, 2, 5]
+    cg_dg_map = np.empty_like(dg_cg_map)
+    for i, im in enumerate(dg_cg_map):
+        cg_dg_map[im] = i
+    # cg_dg_map = [ 0, 7, 8, 1, 5, 9, 3, 6, 4, 2]
+    perms = {}
```

Comment:
This isn't computing a permutation between the DG element and an equivalent CG element, but rather the permutation to a lattice on the cell.

Reply:
Renamed: `dg_to_lattice` and `lattice_to_dg`.